### PR TITLE
userEnabled feature PR #2: Split Tables + Buttons

### PR DESCRIPTION
**Pull request by esarnb**
Split the table into 2 sections: enabled & disabled.
Enable, Disable, and Delete buttons all work. 
Must be authorized as admin to even see the page.

### CODE REVIEW

File: controllers/manage-controller.js [hunk 1]
```diff
@@ -0,0 +1,70 @@
+const { User } = require('../models');
+
+const manageUsers = (req, res) => {
```
Comment (SamMaxwell) [inline controllers/manage-controller.js]: This should be the "getManage" method in users-controller that returns all the users (the route would be GET '/users/manage')

File: controllers/manage-controller.js [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  User
+    .findAll({
+      order: [['name', 'ASC']],
+    //   where: { isEnabled: 0 },
+    })
+    .then((users) => {
```
Comment (SamMaxwell) [inline controllers/manage-controller.js]: the arg could be named generically (e.g. "data")

File: controllers/manage-controller.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      const statuses = users.map((user) => user.get({ plain: true }));
```
Comment (SamMaxwell) [inline controllers/manage-controller.js]: then "statuses" could be named "users"

File: controllers/manage-controller.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+      res.render('users/manage-users', { users: statuses });
+    })
+    .catch(({ message } = {}) => {
+      req.log.error(`User.findAll: ${message}`);
+      res.render('error', { message: 'Users not found' });
```
Comment (SamMaxwell) [inline controllers/manage-controller.js]: every res.render should inject isLoggedIn

File: controllers/manage-controller.js [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+    });
+};
+
+const enableUser = (req, res) => {
```
Comment (SamMaxwell) [inline controllers/manage-controller.js]: I think we could just implement putUser on users-controller (route PUT '/users/123') which could update anything/everything on the user (skills, accomplishments, isEnabled, ...) except for password but it should only modify the the attributes that are actually specified in the body (don't accidentally set a column value to null unintentionally)

File: controllers/manage-controller.js [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    .then(() => {
+      res.status(200).end();
+    })
+    .catch(({ message } = {}) => {
+      req.log.error(`User.update: ${message}`);
+      res.status(500).json({ message: 'could not update disable' });
+    });
+};
+
+const deleteUser = (req, res) => {
```
Comment (SamMaxwell) [inline controllers/manage-controller.js]: could just be named "delete" (if it is in the users-controller) routed to by DELETE /users/123

File: middleware/auth/must-be-admin.js [hunk 7]
```diff
@@ -0,0 +1,9 @@
+const mustBeAdmin = (req, res, next) => {
+  if (!req.session.isAdmin) {
+    res.redirect('/users/my-profile');
```
Comment (SamMaxwell) [inline middleware/auth/must-be-admin.js]: redirect to root '/'

File: public/scripts/users/manage-users.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
+/* eslint-disable no-undef */
+
+const clearErrors = (error) => Object.keys(error)
```
Comment (SamMaxwell) [inline public/scripts/users/manage-users.js]: we should put this into util and re-use it

File: public/scripts/users/manage-users.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+const enableUser = async (event) => {
+  const error = {
+    form: document.querySelector('#manage-error'),
+  };
+
+  clearErrors(error);
+
+  const userid = event.target.dataset.user;
+
+  fetch('/manage/enable', {
```
Comment (SamMaxwell) [inline public/scripts/users/manage-users.js]: why not just use put the id in the url? -> PUT /users/123 { "isDisabled": true }

File: public/scripts/users/manage-users.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+const deleteUser = async (event) => {
+  const error = {
+    form: document.querySelector('#manage-error'),
+  };
+
+  clearErrors(error);
+
+  const userid = event.target.dataset.user;
+
+  fetch('/manage/delete', {
```
Comment (SamMaxwell) [inline public/scripts/users/manage-users.js]: DELETE /users/123

File: public/scripts/users/manage-users.js [hunk 11]
```diff
@@ -0,0 +1,99 @@
...
+  }).then((response) => {
+    if (response.ok) {
+      window.location.reload();
+    }
+  }).catch((err) => {
+    error.form.textContent = err.message;
+  });
+};
+
+const enabledBtns = document.querySelectorAll('.enable-btn');
```
Comment (SamMaxwell) [inline public/scripts/users/manage-users.js]: I think you could add one click event at the document level and check the target's class or something so you don't have to add 100s of event listeners (we had an activity in class regarding this technique)

File: routes/manage/index.js [hunk 12]
```diff
@@ -0,0 +1,11 @@
+const routes = require('express').Router();
```
Comment (SamMaxwell) [inline routes/manage/index.js]: put these in the users-controller

File: views/users/manage-users.handlebars [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+    USERS 
+    <span class="list-group-item-success">Enabled</span> 
+    <span class="list-group-item-danger">Disabled</span>
+    <span>★Admin★</span> 
+    
+    <br />
+    <code id="manage-error"></code>
+    
+    <br />
+    <ol class="list-group list-group-numbered pull-left">
```
Comment (SamMaxwell) [inline views/users/manage-users.handlebars:11]: maybe pass two lists in: disabledUsers and enabledUsers?

File: views/users/manage-users.handlebars [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+        {{#each users as |user| }}
+            <span>
```
Comment (SamMaxwell) [inline views/users/manage-users.handlebars]: do you need the <span> for some reason?

File: views/users/manage-users.handlebars [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+            {{#if isEnabled}}
+             <li class="list-group-item d-flex justify-content-between list-group-item-success">
+                    <h3 class="mt-3"> {{name}} {{#if isAdmin}} ★ {{/if}} 
+                        <button type="button" class="btn btn-danger pull-right delete-btn" data-bs-toggle="tooltip" data-bs-placement="top" title="Remove from DB" data-user={{id}}>Delete</button>
+                        <button type="button" class="btn btn-secondary pull-right disable-btn" data-bs-toggle="tooltip" data-bs-placement="top" title="Revoke Public Access" data-user={{id}}>Disable</button>
+                    </h3>
+            </li>
+            {{/if}}
+            </span>
+        </li>
```
Comment (SamMaxwell) [inline views/users/manage-users.handlebars]: this doesn't seem to belong here. is it extra?

File: views/users/manage-users.handlebars [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+        {{/each}}
+    </ol>
+    <ol class="list-group list-group-numbered pull-right">
+        {{#each users as |user| }}
+            <span>
```
Comment (SamMaxwell) [inline views/users/manage-users.handlebars]: ditto

File: views/users/manage-users.handlebars [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+            {{#unless isEnabled}}
+            <li class="list-group-item d-flex justify-content-between list-group-item-danger">
+                    <h3 class="mt-3"> {{name}} {{#if isAdmin}} ★ {{/if}} 
+                        <button type="button" class="btn btn-danger pull-right delete-btn" data-bs-toggle="tooltip" data-bs-placement="top" title="Remove from DB" data-user={{id}}>Delete</button>
+                        <button type="button" class="btn btn-success pull-right enable-btn" data-bs-toggle="tooltip" data-bs-placement="top" title="Add Public Access" data-user={{id}}>Enable</button>
+                    </h3>
+            </li>
+            {{/unless}}
+            </span>
+        </li>
```
Comment (SamMaxwell) [inline views/users/manage-users.handlebars]: ditto

**Review by SamMaxwell — CHANGES_REQUESTED** [review]

File: controllers/manage-controller.js (hunk 1 above)
Comment (esarnb) [reply controllers/manage-controller.js]: sure, renamed to getManage and will be put into controllers/users-controller

File: controllers/manage-controller.js (hunk 2 above)
Comment (esarnb) [reply controllers/manage-controller.js]: sure, renamed to data and passed to variable renamed to users

File: controllers/manage-controller.js (hunk 4 above)
Comment (esarnb) [reply controllers/manage-controller.js]: Added isLoggedIn to the render manage and error

File: controllers/manage-controller.js (hunk 5 above)
Comment (esarnb) [reply controllers/manage-controller.js]: I would assume this means like `User.update(req,body, {where: { id: req.body.userid } })`.
So like req.body is defined by whatever made the put request. 
I also added a if statement to send back a res 405 for a field editing password.

File: controllers/manage-controller.js (hunk 6 above)
Comment (esarnb) [reply controllers/manage-controller.js]: Moved to user controller and renamed to delete

File: public/scripts/users/manage-users.js (hunk 8 above)
Comment (esarnb) [reply public/scripts/users/manage-users.js]: Added to utils and imported to public js files

File: controllers/manage-controller.js (hunk 6 above)
Comment (esarnb) [reply controllers/manage-controller.js]: Actually in the IDE the word `delete` is reserved, so I cannot run it unless renamed to deleteUser or similar

File: public/scripts/users/manage-users.js (hunk 11 above)
Comment (esarnb) [reply public/scripts/users/manage-users.js]: I now realize we have Jquery, i will utilize event delegation

File: views/users/manage-users.handlebars (hunk 15 above)
Comment (esarnb) [reply views/users/manage-users.handlebars]: yes thank you for finding this

File: views/users/manage-users.handlebars (hunk 14 above)
Comment (esarnb) [reply views/users/manage-users.handlebars]: it was leftover from trying to push buttons to the right, but `pull-right` from bootstrap fixed many things!
So i dont need these spans

File: public/scripts/users/manage-users.js (hunk 8 above)
Comment (esarnb) [reply public/scripts/users/manage-users.js]: public/script/js cannot recognize `require` if i try to import utils. Maybe we can have a utils within the public folder too?

**Review by esarnb — COMMENTED** [review]
I have made my changes in my IDE, so that I can test and run the app.
I am going to make a fixes commit, then make a new pull request

**Pull request closed without merging**